From 26261361fdbf699e19020f39568eedd91c710184 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Wed, 8 Nov 2023 15:59:20 +0100
Subject: [PATCH 858/985] evl/net: socket: introduce protocol-specific bits as
 union

Plan for adding more protocol-specific data, such as ipv4 bits. Let's
have them grouped into a union in the EVL socket descriptoor.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/net/socket.h       | 14 ++++++++------
 kernel/evl/net/packet/packet.c | 18 +++++++++---------
 2 files changed, 17 insertions(+), 15 deletions(-)

diff --git a/include/evl/net/socket.h b/include/evl/net/socket.h
index 8b4b5a273..4ec2116f7 100644
--- a/include/evl/net/socket.h
+++ b/include/evl/net/socket.h
@@ -71,12 +71,14 @@ struct evl_socket {
 	struct evl_wait_queue wmem_wait;
 	struct evl_crossing wmem_drain;
 	int protocol;
-	struct {
-		int real_ifindex;
-		int vlan_ifindex;
-		u16 vlan_id;
-		u32 proto_hash;
-	} binding;
+	union {
+		struct {
+			int real_ifindex;
+			int vlan_ifindex;
+			u16 vlan_id;
+			u32 proto_hash;
+		} packet;
+	} u;
 	hard_spinlock_t oob_lock;
 };
 
diff --git a/kernel/evl/net/packet/packet.c b/kernel/evl/net/packet/packet.c
index cd765ea29..86ff952dc 100644
--- a/kernel/evl/net/packet/packet.c
+++ b/kernel/evl/net/packet/packet.c
@@ -74,11 +74,11 @@ static bool __packet_deliver(struct evl_net_rxqueue *rxq,
 		 * a newly registered device reusing an old ifindex we
 		 * initially captured at binding time.
 		 */
-		ifindex = READ_ONCE(esk->binding.real_ifindex);
+		ifindex = READ_ONCE(esk->u.packet.real_ifindex);
 		if (ifindex) {
 			if (ifindex != dev->ifindex)
 				continue;
-			vlan_id = READ_ONCE(esk->binding.vlan_id);
+			vlan_id = READ_ONCE(esk->u.packet.vlan_id);
 			if (skb_vlan_tag_get_id(skb) != vlan_id)
 				continue;
 		}
@@ -218,7 +218,7 @@ static int attach_packet_socket(struct evl_socket *esk,
 	evl_spin_lock_irqsave(&protocol_lock, flags);
 
 	esk->proto = proto;
-	esk->binding.proto_hash = hkey;
+	esk->u.packet.proto_hash = hkey;
 	esk->protocol = protocol;
 
 	_rxq = find_rxqueue(hkey);
@@ -251,7 +251,7 @@ static void detach_packet_socket(struct evl_socket *esk)
 
 	evl_spin_lock_irqsave(&protocol_lock, flags);
 
-	rxq = find_rxqueue(esk->binding.proto_hash);
+	rxq = find_rxqueue(esk->u.packet.proto_hash);
 
 	list_del_init(&esk->next_sub); /* Remove from rxq->subscribers */
 	if (list_empty(&rxq->subscribers)) {
@@ -292,7 +292,7 @@ static int bind_packet_socket(struct evl_socket *esk,
 
 	mutex_lock(&esk->lock);
 
-	old_ifindex = esk->binding.vlan_ifindex;
+	old_ifindex = esk->u.packet.vlan_ifindex;
 	if (new_ifindex != old_ifindex) {
 		if (new_ifindex) {
 			/* @dev has to be a VLAN device. */
@@ -339,9 +339,9 @@ static int bind_packet_socket(struct evl_socket *esk,
 	raw_spin_lock_irqsave(&esk->oob_lock, flags);
 	if (new_ifindex != old_ifindex) {
 		/* First change the real interface, next the vid. */
-		WRITE_ONCE(esk->binding.real_ifindex, real_ifindex);
-		esk->binding.vlan_id = vlan_id;
-		WRITE_ONCE(esk->binding.vlan_ifindex, new_ifindex);
+		WRITE_ONCE(esk->u.packet.real_ifindex, real_ifindex);
+		esk->u.packet.vlan_id = vlan_id;
+		WRITE_ONCE(esk->u.packet.vlan_ifindex, new_ifindex);
 	}
 	raw_spin_unlock_irqrestore(&esk->oob_lock, flags);
 
@@ -357,7 +357,7 @@ static int bind_packet_socket(struct evl_socket *esk,
 static struct net_device *get_netif_packet(struct evl_socket *esk)
 {
 	return  evl_net_get_dev_by_index(esk->net,
-					esk->binding.vlan_ifindex);
+					esk->u.packet.vlan_ifindex);
 }
 
 static struct net_device *find_xmit_device(struct evl_socket *esk,
-- 
2.30.2

